src/wrapper/jnienv.rs

```diff
@@ -1973,20 +1986,46 @@ impl<'a> JNIEnv<'a> {
         jni_error_code_to_result(res)
     }
 
-    /// Return a tuple with a pointer to elements of the given Java byte array as first element.
+    /// Return a tuple with a pointer to elements of the given Java array as first element.
     /// The tuple's second element indicates if the pointed-to array is a copy or not.
     ///
-    /// The result is valid until the corresponding release_byte_array_elements() function is
+    /// The result is valid until the corresponding release_array_elements() function is
     /// called. Since the returned array may be a copy of the Java array, changes made to the
     /// returned array will not necessarily be reflected in the original array until
-    /// release_byte_array_elements() is called.
+    /// release_array_elements() is called.
+    ///
+    /// To determine the type, type annotation can be required on the return value:
+    /// let (ptr, is_copy): (*mut jbyte, _) = env.get_array_elements(java_array)?;
+    /// If preferred, this method can be invoked using explicit type annotation:
+    /// let (ptr, is_copy) = env.get_array_elements::<jbyte>(java_array)?;
     ///
-    /// See also [`release_byte_array_elements`](struct.JNIEnv.html#method.release_byte_array_elements)
-    pub fn get_byte_array_elements(&self, array: jbyteArray) -> Result<(*mut jbyte, bool)> {
-        non_null!(array, "get_byte_array_elements array argument");
+    /// See also [`release_array_elements`](struct.JNIEnv.html#method.release_array_elements)
+    pub fn get_array_elements<T: JType>(&self, array: jarray) -> Result<(*mut T, bool)> {
+        non_null!(array, "get_array_elements array argument");
         let mut is_copy: jboolean = 0xff;
-        let ptr = jni_non_void_call!(self.internal, GetByteArrayElements, array, &mut is_copy);
-        Ok((ptr, is_copy == sys::JNI_TRUE))
+        let type_name = type_name::<T>();
```

Comment:
According to the [documentation](https://doc.rust-lang.org/std/any/fn.type_name.html), type name representation isn't specified, so we shouldn't rely on the specific values.

Reply:
I agree. There's also `TypeId::of::<T>()`, which is more robust (types are compared to types). But, it cannot be used in a `match`... we're left with a chain of `if`s, which is anything but elegant.

The only options I can think of are:
- Put the version with the chain of `if`s into a macro, and try to use it in different parts of the code base. I'll push that version, so you can review it.
- Don't do anything, and resort to writing explicit code on demand (!).

Reply:
OK, I've removed the trait, and switched to `TypeId::of`. Please take a look at [the last commit](https://github.com/jni-rs/jni-rs/pull/273/commits/d119785ab29d02fcceec9fd06d87d980fd65d557).

Reply:
OK, I've implemented `AutoArray<T>`, and used it to implement `get_auto_byte_array()`. Please take a look when you can.